Path resolution docs (current state only) v2 (#11442)
ethereum/solidity: docs/layout-of-source-files.rst

  import {symbol1 as alias, symbol2} from "filename";

.. index:: virtual filesystem, source unit name, import; path, filesystem path

Import paths
------------

> The paths used in imports in a general case do not have to be filesystem paths.

Reviewer: Maybe we could also define that an import path is the parameter used within an import statement, `import "<import-path>"`.

Author: I added a sentence that says what an import path is in the "Syntax and Semantics" section above.

I have also rewritten the "Import paths" section here to say more about relation between an import path and a source unit name.

The compiler maintains an internal database (:ref:`virtual filesystem <virtual-filesystem>`) where
each compiled source unit is assigned a unique *source unit name* which is an opaque and unstructured identifier.
The import path is translated into a source unit name and used to find the corresponding source unit
in this database.

For a complete description of the virtual filesystem and the path resolution logic used by the

> compiler see :ref:`Path Resolution <path-resolution>`.

Author: Maybe I should put the docs back here now that they are shorter?

They're still long enough to warrant a separate page so I left them as a separate page for now but I can easily change it.

  import {symbol1 as alias, symbol2} from "filename";

.. index:: virtual filesystem, source unit name, import; path, filesystem path, import callback, Remix IDE

> Import paths

Reviewer 2: ```suggestion
Import Paths
```

------------

In order to be able to support reproducible builds on all platforms, the Solidity compiler has to
abstract away the details of the filesystem where source files are stored.
For this reason import paths do not refer directly to files in the host filesystem.
Instead the compiler maintains an internal database (*virtual filesystem* or *VFS* for short) where
each source unit is assigned a unique *source unit name* which is an opaque and unstructured identifier.
The import path specified in an import statement is translated into a source unit name and used to
find the corresponding source unit in this database.

> Using the :ref:`Standard JSON <compiler-api>` API it is possible to directly provide the content of
> all the source files as a part of compiler's input.

Reviewer 2: ```suggestion
Using the :ref:`Standard JSON <compiler-api>` API it is possible to directly provide the names and contents of
all the source files as a part of the compiler input.
```

In this case source unit names are truly arbitrary.

> If, however, you want the compiler to automatically find and load source code into the VFS, your

Reviewer 2: ```suggestion
If, however, you want the compiler to automatically find and load source code from the host filesystem into the VFS, your
```

Author: It's worded it in a general way on purpose. This sentence is about the generic import callback and it might load content from a different source (github, npm registry, etc.).

source unit names need to be structured in a way that makes it possible for an :ref:`import callback
<import-callback>` to locate them.
When using the command-line compiler the default import callback supports only loading source code
from in the host filesystem, which means that your source unit names must be paths.

> Some environments, provide custom callbacks that are more versatile.

Reviewer 2: ```suggestion
Some environments provide custom callbacks that are more versatile.
```
